### PR TITLE
Fix implicit fallthrough warning and unused variable warning in GCC

### PR DESCRIPTION
**Pull request by jackgerrits**

### CODE REVIEW

File: vowpalwabbit/gen_cs_example.h [hunk 1]
```diff
@@ -144,7 +144,7 @@ void gen_cs_label(cb_to_cs& c, example& ec, COST_SENSITIVE::label& cs_ld, uint32
 }
 
 template <bool is_learn>
-void gen_cs_example_dr(cb_to_cs& c, example& ec, CB::label& ld, COST_SENSITIVE::label& cs_ld, float clip_p=0.f)
+void gen_cs_example_dr(cb_to_cs& c, example& ec, CB::label& ld, COST_SENSITIVE::label& cs_ld, float /*clip_p*/ =0.f)
```
Comment (lokitoth) [inline vowpalwabbit/gen_cs_example.h]: Does the default value still need to be here even though the variable is unused?

File: explore/hash.h [hunk 2]
```diff
@@ -112,14 +112,20 @@ inline uint64_t uniform_hash(const void * key, size_t len, uint64_t seed)
 
   switch (len & 3)
   {
-  case 3: k1 ^= tail[2] << 16;
-  case 2: k1 ^= tail[1] << 8;
+  case 3:
+    k1 ^= tail[2] << 16;
+    // fall through
```
Comment (lokitoth) [inline explore/hash.h]: Does the presence of the comment suppress the warning, or does the command have to be exactly `// fall through`? If not the latter, could you add a note about this suppressing the warning, so it does not get removed later?

Comment (jackgerrits) [reply explore/hash.h]: Yep! GCC see's this as an annotation and silences the warning. There is also a gnu specific attribute available. In C++17 a [[fallthrough]] attribute was added for this exact purpose. 

I figured a comment is simple and suffices for now. I'll add a description so we don't accidentally remove it.

File: vowpalwabbit/gen_cs_example.h (hunk 1 above)
Comment (jackgerrits) [reply vowpalwabbit/gen_cs_example.h]: Yes, it allows the function to be called without the last parameter (as is done in existing code). If I remove it the last parameter becomes required.

Comment (jackgerrits) [reply vowpalwabbit/gen_cs_example.h]: Yes, it allows the function to be called without the last parameter (as is done in existing code). If I remove it the last parameter becomes required.

Comment (lokitoth) [reply vowpalwabbit/gen_cs_example.h]: Makes sense.

**Pull request merged**